RileyLinkKit/PumpOpsSynchronous.swift

```diff
@@ -483,38 +421,73 @@ class PumpOpsSynchronous {
             }
 
             let page = try HistoryPage(pageData: pageData, pumpModel: pumpModel)
+            
+            let (timeStampedEvents, hasMoreEvents) = convertPumpEventToTimestampedEvents(pumpEvents: page.events.reversed(), startDate: startDate, pumpModel: pumpModel)
+            
+            events.append(contentsOf: timeStampedEvents)
+            
+            if !hasMoreEvents {
+                break
+            }
+        }
+        return (events, pumpModel)
+    }
 
-            for event in page.events.reversed() {
-                if let event = event as? TimestampedPumpEvent, !seenEventData.contains(event.rawData) {
-                    seenEventData.insert(event.rawData)
-
-                    var timestamp = event.timestamp
-                    timestamp.timeZone = pump.timeZone
-
-                    if let date = timestamp.date?.addingTimeInterval(timeAdjustmentInterval) {
-                        if date.timeIntervalSince(startDate) < -eventTimestampDeltaAllowance {
-                            NSLog("Found event at (%@) to be more than %@s before startDate(%@)", date as NSDate, String(describing: eventTimestampDeltaAllowance), startDate as NSDate);
-                            break pages
-                        } else if date.timeIntervalSince(timeCursor) > eventTimestampDeltaAllowance {
-                            NSLog("Found event (%@) out of order in history. Ending history fetch.", date as NSDate)
-                            break pages
-                        } else {
-                            if (date.compare(startDate) != .orderedAscending) {
-                                timeCursor = date
-                            }
-                            events.insert(TimestampedHistoryEvent(pumpEvent: event, date: date), at: 0)
+    ///
+    /// Converts PumpEvents after startDate to TimestampedHistoryEvents
+    /// hasMoreEvents indicates the caller should continue getting more PumpEvents from the device
+    ///
+    internal func convertPumpEventToTimestampedEvents(pumpEvents: [PumpEvent], startDate: Date, checkDate: Date = Date(), pumpModel: PumpModel) -> (events: [TimestampedHistoryEvent], hasMoreEvents: Bool) {
```

Comment:
checkDate is not used.  I think I know why you have it here: because the planned change is to call isMutable(at:) which requires a date.  I think we'll have another check instead of isMutable, that doesn't take a date, as our check for history fetching is not time-dependent.